File: ds/min-stack/sol-1.js
/**
 * initialize your data structure here.
 */
var MinStack = function () {
  this.head = null;
  this.tail = null; // ***
  this.size = 0;
};

// constructor function -> which is nothing but class
var Node = function (value) {
  this.value = value;
  this.next = null;
};

MinStack.prototype.toString = function () {
  return `Current Head : ${this.head.value}, next head : ${this.head.next.value}`;
};

/**
 * @param {number} x
 * @return {void}
 */
MinStack.prototype.push = function (x) {
  // !PUSH -> Adding the element @ the top of the stack
  // ! NOTE: Array.prototype.push() adds the element @ end of the array
  let newNode = new Node(x);

  if (!this.head) {
    // new list then head and tail points at this newly added node
    this.head = this.tail = newNode;
  } else {
    // linkedList => [2, 5, 3] i.e-  (this.head) 2 -> 5 -> 3 (this.tail)
    // want to push 6 => [6, 2, 5, 3] steps:
    let currentNode = this.head; // currentNode = 2 = this.head
    this.head = newNode; // (this.head) 6       (currentNode) 2 -> 5 -> 3 (this.tail)
    this.head.next = currentNode; // Now provide link/reference of newly added node to the list using:- next
    // (this.head) 6 -> 2 -> 5 -> 3 (this.tail)
  }

  this.size++;
};

/**
 * @return {void}
 */

MinStack.prototype.pop = function () {
  // !POP ->  Removing the element from the top of the stack
  // ! NOTE: Array.prototype.pop() -> removes the element form the last of array

  if (this.size === 1) {
    this.head = this.tail = null;
  } else {
    // list [6, 2, 5, 3], need to pop first element i.e: 6
    // (this.head) 6 -> 2 -> 5 -> 3 (this.tail)
    let nextReference = this.head.next; // nextReference = 2 (2nd Node)
    this.head = nextReference; // de-referencing the current head and re-referencing it to 2nd Node (next node)
    // 6 {Node got de-referenced}{lost in universe}      (this.head) 2 -> 5 -> 3 (this.tail)
  }
  this.size--;
};

/**
 * @return {number}
 */
MinStack.prototype.top = function () {
  return this.head.value;
};

/**
 * @return {number}
 */
MinStack.prototype.getMin = function () {
  let minValue = this.head.value; // let assume head value is min value
  let cloneHead = this.head;
  for (let i = 0; i < this.size - 1; i++) {
    cloneHead = cloneHead.next; // cloneHead -> Type: Node  and  cloneHead.next -> Type: Node
    if (cloneHead.value < minValue) {
      minValue = cloneHead.value;
    }
  }

  return minValue;
};

/**
 * Your MinStack object will be instantiated and called as such:
 * var obj = new MinStack()
 * obj.push(x)
 * obj.pop()
 * var param_3 = obj.top()
 * var param_4 = obj.getMin()
 */

// !NOTE: Push and Pop operations happens on the top/begining element only

let obj = new MinStack();
// [6, 2, 5, 3]
obj.push(3);
obj.push(5);
console.log(obj.toString());
obj.push(2);
console.log(obj.toString());
obj.push(6);
console.log(obj.toString());
console.log('min value', obj.getMin());

obj.pop();
console.log('After poping the value -> ', obj.toString());

console.log('top Value', obj.top());

File: algo/LRU-cache/sol-1.java
class LRUCache {
    
    final Node head=new Node();
    final Node tail=new Node();
    Map<Integer, Node> node_map;
    int cache_capacity;
    
    public LRUCache(int capacity) {
        node_map=new HashMap(capacity);
        this.cache_capacity=capacity;
        head.next=tail;
        tail.prev=head;
    }
    
    public int get(int key) {
        int result=-1;
        Node node = node_map.get(key);
        if(node!=null){
            result=node.val;
            remove(node);
            add(node);
        }
        return result;
    }
    
    public void put(int key, int value) {
        Node node=node_map.get(key);
        if(node!=null){
            remove(node);
            node.val=value;
            node.key=key;
            add(node);
        }else{
            if(node_map.size() == cache_capacity){
                node_map.remove(tail.prev.key);
                remove(tail.prev);
            }
            
            Node new_node = new Node();
            new_node.key=key;
            new_node.val=value;
            node_map.put(key,new_node);
            add(new_node);
        }
    }
    
    public void add(Node node) {
        Node head_next=head.next;
        node.next=head_next;
        head_next.prev=node;
        head.next=node;
        node.prev=head;
    }
    
    public void remove(Node node) {
        Node next_node=node.next;
        Node prev_node=node.prev;
        
        next_node.prev=prev_node;        
        prev_node.next=next_node;
    }
    
    class Node {
        int key;
        int val;
        Node prev; 
        Node next;
    }
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */